File: standard/merge_sort.cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;


// recursive divide array and merge 


void merge(vector<int> &arr, int l, int mid, int r){

    int temp[r-l+1];

    int i=l, j=mid+1, k=0;

    while(i<=mid && j<=r){

        if(arr[i]<arr[j]) {temp[k++] = arr[i++]; /* i++; */ }
        else {temp[k++] = arr[j++]; /* j++;*/}
    
        // k++;
    }

    while(i<=mid){
        temp[k++] = arr[i++];
        // k+=1;
        // i+=1;
    }


    while(j<=r){
        temp[k++] = arr[j++];
        // k+=1;
        // j+=1;
    }

    for(int t=l; t<=r; t++){
        arr[t] = temp[t-l];
    }

}


void merge_sort(vector<int> &arr, int l, int r){


    if(l<r){
        int mid = l + (r-l)/2;

        merge_sort(arr, l, mid);
        merge_sort(arr, mid+1, r);
    
        merge(arr, l, mid, r);
    }

}



int main(){
    
    vector<int> test = {10,9,11,7,6,5,4,3,2,1};

    for(int k=0; k<test.size(); k++){
        test[k] = (rand()%100)+test[k];
    }

    merge_sort(test, 0, test.size()-1);

    for(auto i : test) cout<<i<<" ";
    
    return 0;
}

File: misc/MergeSort.java
import java.util.*;

class MergeSort {
    public static void main(String ...args){
        int[] arr = {2,34,17,19,63,37,10,51,26,3,11,36,5,3,72};

        System.out.println(Arrays.toString(arr));

        new MergeSort().mergeSort(arr, 0, arr.length-1);

        System.out.println(" ---------- ");
        System.out.println(Arrays.toString(arr));

    }

    void mergeSort(int[] arr, int l, int r){
        if(l >= r ){
            return ;
        }
        int m = (r-l)/2 + l;

        mergeSort(arr, l, m);
        mergeSort(arr, m+1, r);

        merge(arr, l, r);
    }

    void merge(int[] arr, int l, int r){
        int m = (r-l)/2 + l;
        int i=l, j=m+1, k=0;
        
        System.out.print("\nl:"+l+" r:"+r+" temp:"+(r-l+1));

        int temp[] = new int[r-l+1];

        while(i<=m && j<=r){
            if(arr[i]<arr[j]) temp[k++] = arr[i++];
            else temp[k++] = arr[j++];
        }

        while(i<=m){
            temp[k++] = arr[i++];
        }

        while(j<=r){
            temp[k++] = arr[j++];
        }

        System.out.print("{ ");
        for(i=l, k=0; k<temp.length; i++, k++){
            arr[i] = temp[k];
            System.out.print(temp[k]+", ");
        }
        System.out.print(" }\n");
    }
    
}